### ISSUE TITLE
wrong overload resolved for `torch.mul(x, 4)` in `__torch_dispatch__`

### ISSUE THREAD

**Issue by makslevental** [pytorch/pytorch#77332]
### 🐛 Describe the bug

minimal repro

```python
import torch
class ElementwiseMulScalarIntModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return torch.mul(x, 4)

class TrivTensor(torch.Tensor):
    @staticmethod
    def __new__(cls, elem, *, requires_grad=None):
        if requires_grad is None:
            return super().__new__(cls, elem)
        else:
            return cls._make_subclass(cls, elem, requires_grad)

    @classmethod
    def __torch_dispatch__(cls, func, _types, args=(), kwargs=None):
        print(func)
        for arg in args:
            if isinstance(arg, cls): continue
            print(arg, type(arg))


ElementwiseMulScalarIntModule()(TrivTensor(torch.randint(10, (3, 4))))

> aten.mul.Tensor
> 4 <class 'int'>
```

might be related to https://github.com/pytorch/pytorch/issues/77223

cc @silvasean @cathyzhyi @powderluv 

### Versions

PyTorch version: 1.12.0.dev20220511+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.21.2
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.13.0-40-generic-x86_64-with-glibc2.29
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 510.39.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.12.0.dev20220511+cpu
[pip3] torchvision==0.13.0.dev20220511+cpu
[conda] No relevant packages

**Comment by makslevental** [comment]
okay I understand thanks. the thing is I don't know where to "place blame" in this situation:

```python
import torch
from torch.fx.experimental.proxy_tensor import no_dispatch
from torch.fx.operator_schemas import normalize_function
class ElementwiseMulScalarIntModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return torch.mul(x, 4)

class TrivTensor(torch.Tensor):
    @staticmethod
    def __new__(cls, elem, *, requires_grad=None):
        if requires_grad is None:
            return super().__new__(cls, elem)
        else:
            return cls._make_subclass(cls, elem, requires_grad)

    @classmethod
    def __torch_dispatch__(cls, func, _types, args=(), kwargs=None):
        with no_dispatch():
            assert isinstance(args[0], torch.Tensor)
            assert isinstance(args[1], int)
            assert func == torch.ops.aten.mul.Tensor
            new_args_and_kwargs = normalize_function(func, args, {}, (torch.Tensor, int), {},
                normalize_to_only_use_kwargs=True,
            )
            _, new_kwargs = new_args_and_kwargs
            assert isinstance(new_kwargs["input"], cls)
            assert isinstance(new_kwargs["input"], torch.Tensor)
            assert isinstance(new_kwargs["other"], int)

ElementwiseMulScalarIntModule()(TrivTensor(torch.ones(10)))
```

So inside `__torch_dispatch__` you get the `Tensor` overload and an integer `other` argument and then `normalize_function` doesn't complain (because [`candidate_signature.bind`](https://github.com/pytorch/pytorch/blob/6066e5929f12cdef52fe5d85af8001ad3583ec42/torch/fx/operator_schemas.py#L291) succeeds) but also doesn't perform the type promotion. You could say "garbage in -> garbage out" wrt `normalize_function`, or that it's not responsible for type promotion, but ultimately I simply don't know how to resolve the correct signature for the op.

**Comment by ezyang** [comment]
IMO it's normalize's fault but I am also not super familiar with that code
